addons/mail/models/mail_activity.py

```diff
@@ -735,6 +736,67 @@ def _compute_activity_state(self):
             else:
                 record.activity_state = False
 
+    def _search_activity_state(self, operator, value):
+        all_states = {'overdue', 'today', 'planned', False}
+        if operator == '=':
+            states = {value}
+        elif operator == '!=':
+            states = all_states - {value}
+        elif operator == 'in':
+            states = set(value)
+        elif operator == 'not in':
+            states = all_states - set(value)
+
+        if False in states:
+            # If we search "activity_state = False", they might be a lot of records
+            # (million for some models), so instead of returning the list of ID
+            # [(id, 'in', ids)] we will reverse the domain and return something like
+            # [(id, 'not in', ids)], so the list of ids is as small as possible
+            reverse_domain = True
+            states = all_states - states
+        else:
+            reverse_domain = False
+
+        # Use number in the SQL query for performance purpose
+        integer_state_value = {
+            'overdue': -1,
+            'today': 0,
+            'planned': 1,
+            False: None,
+        }
+
+        states = {integer_state_value.get(s or False) for s in states}
```

Comment:
Maybe rename to states_int (using the same variable for two different content does not help understanding).